Generate formulas automatically from the component graph

shsms opened the pull request:
Closes #102

shsms committed:
Make inverter type available through the component graph

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Update mock api to support inverter types

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Update `ResampledFormulaBuilder` to be a subclass of `FormulaBuilder`

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add `FormulaGenerator` class and grid_power/battery_power formulas

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms commented:
This PR currently implements just one `grid_power` formula.  I will add more in new commits,  but the other parts are already ready for a first look.

ela-kotulska-frequenz commented on src/frequenz/sdk/microgrid/component/_component.py:
```diff
@@ -62,6 +95,7 @@ class Component:
 
     component_id: int
     category: ComponentCategory
+    type: ComponentType = InverterType.NONE
```
It is missleading that Category type has InverterType as default.
Should we have any default at all?
Do you plan to add type for all components in this PR and then remove this default?
If yes, then please comment it for easier PR :)

shsms replied:
Yes,  I'd like to make it not default,  but `Component` objects are getting created literally everywhere in the codebase,  and so many in the tests,  so removing the default will be a huge change.  I already tried,  but gave up (temporarily) after around 500 lines.

Don't want to add types of other components yet either,  won't work for now,  the microgrid api needs an update for this to work.
 
I will try and make it an `Optional` field,  with `None` as default value.  That should work I think.

shsms replied:
This is done.

ela-kotulska-frequenz commented on src/frequenz/sdk/timeseries/logical_meter/_resampled_formula_builder.py:
```diff
@@ -59,7 +59,7 @@ async def _get_resampled_receiver(self, component_id: int) -> Receiver[Sample]:
 
     async def push_component_metric(
-        self, component_id: int, nones_are_zeros: bool
+        self, component_id: int, *, nones_are_zeros: bool
```
What * mean? It is not described.

shsms replied:
It is part of the Python syntax.  just means that all the arguments after the `*` are keyword-only i.e., they can only be passed with the name, like `push_component_metric(10, nones_are_zeros = False)`.

https://peps.python.org/pep-3102/

ela-kotulska-frequenz commented on src/frequenz/sdk/timeseries/logical_meter/_logical_meter.py:
```diff
@@ -107,3 +114,27 @@ async def start_formula(
+        )
+        return out_chan.new_receiver()
+
+    # async def battery_power(self) -> Receiver[Sample]:
```
Unnecessary comment?

shsms replied:
It was supposed to be for the next commit.  I've moved it now.

ela-kotulska-frequenz commented:
For me looks good.
Waiting for more formulas to continue review :)

leandro-lucarella-frequenz commented on src/frequenz/sdk/microgrid/component/_component.py:
```diff
@@ -8,9 +8,42 @@
 
 
+class ComponentType(Enum):
+    """A ABC from which individual component types are derived."""
```
This is an enum, not a ABC.

shsms replied:
Fixed.

leandro-lucarella-frequenz commented on src/frequenz/sdk/microgrid/component/_component.py:
```diff
@@ -8,9 +8,42 @@
+    """A ABC from which individual component types are derived."""
+
+
+class InverterType(ComponentType):
```
Any reason to derive from an empty `Enum`?

shsms replied:
Just for typing.  Then we can have a single `type` field in the `Component` object, and assign `InverterType` or `MeterType` values to it and mypy will still be happy.  That way we'd also get an interface that follows the protobuf definition somewhat.

leandro-lucarella-frequenz replied:
~But what I don't understand is, if you are using a method that returns `ComponentType`, then you can't access any of the `InverterType` enum members, right? This is the part I don't understand~ Never mind :man_facepalming:

shsms replied:
it doesn't return a `ComponentType`.  it returns an `InverterType` or `MeterType`, etc.  The type hint says `ComponentType`  and that keeps mypy happy,  because the return values do implement `ComponentType` (because it is empty).

leandro-lucarella-frequenz commented on src/frequenz/sdk/microgrid/component/_component.py:
```diff
@@ -8,9 +8,42 @@
+
+def _component_type_from_protobuf(
+    component_type: inverter_pb.Type.ValueType,
+) -> ComponentType:
```
Am I crazy or the user of this function will have to forcibly cast the return to `InverterType` to be of any use?

shsms replied:
A typehint from a super class doesn't change the type of the object being returned, in Python.  So the function still returns `InverterType.BATTERY` values that's directly comparable.  You're forgetting that we're using Python,  so yes, you are crazy \o/

leandro-lucarella-frequenz replied:
Ah, no, I remember this is Python, but the type checker is still a type checker and it won't let you do this:

```py
ct = _component_type_from_protobuf(...)
if ct == ct.BATTERY:
    ...
```

for example, but of course this is not what you want, the trick is you are still always comparing with the specialized type member:

```py
ct = _component_type_from_protobuf(...)
if ct == InverterType.BATTERY:
    ...
```

So you don't need to access any members in the return type, just compare it.

So yeah, all good, sorry about the brain fart :dash: 🤭

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/logical_meter/_formula_generators.py:
```diff
@@ -0,0 +1,93 @@
+
+    def __init__(
+        self,
+        namespace: str,
```
Maybe `request_namespace` or something that makes it more obvious what is referring to? Not sure, just a suggestion.

shsms replied:
I think we need to qualify it only when we move out of the data pipeline.  It has a shared meaning in the data pipeline code,  so I'd say this is fine.

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/logical_meter/_formula_generators.py:
```diff
@@ -0,0 +1,93 @@
+            if comp.category == ComponentCategory.INVERTER:
+                nones_are_zeros = True
+            elif comp.category == ComponentCategory.METER:
+                nones_are_zeros = False
```
Maybe explain why wants `None`s as zeros and the other doesn't?

shsms replied:
I've updated the comment.

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/logical_meter/_formula_generators.py:
```diff
@@ -0,0 +1,93 @@
+        self._resampler_subscription_sender = resampler_subscription_sender
+        self._namespace = namespace
+
+    async def grid_power_formula(
```
Is the idea to keep adding formulas to this `FormulaGenerator`? If so I would recommend to go for a design that allows us to add more formulas without having all of them in the same file/class, as it could grow very big, as it happened with `MicrogridData`. Maybe having one class per formula generator instead?

shsms replied:
That makes sense,  I'm working on it.

shsms replied:
done

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/logical_meter/_logical_meter.py:
```diff
@@ -32,6 +37,7 @@ def __init__(
         self,
         channel_registry: ChannelRegistry,
         resampler_subscription_sender: Sender[ComponentMetricRequest],
+        component_graph: ComponentGraph,
```
Shouldn't this be taken from the `microgrid.get()` singleton so it can pick up updates to the component graph? In any case it is fine for me to leave it like this for now, but I'm asking more to think about how we want things to work in the future, to make sure we all point at the same direction.

In any case, using `microgrid.get().component_graph` will not also help if the component graph is updated, because then the formula should also be regenerated. I'm starting to think that we need an actor or something that can notify (to whoever is listening) about component graph changes.

shsms replied:
We can also subscribe to config changes from the component graph.  I guess we'll need a sprint to update all parts of the sdk once cloud sync starts to support the sdk,  and we start using the config manager.

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/logical_meter/_logical_meter.py:
```diff
@@ -107,3 +117,32 @@ async def start_formula(
+        )
+        return out_chan.new_receiver()
+
+    async def grid_power(self) -> Receiver[Sample]:
```
I would try to make it more clear that this is actually creating a new receiver. At the method name level I think it could be difficult without making it look **really** ugly, but at least it should be pointed out very clearly in the docs.

shsms replied:
Updated the docstring.

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/logical_meter/_formula_generators.py:
```diff
@@ -91,3 +91,61 @@ async def grid_power_formula(
+        that are attached to batteries.
+
+        If there's no data coming from an inverter, that inverter's power will be
+        treated as 0.
```
This will be moved to the batter pool eventually, right?

shsms replied:
No,  the battery *power* formula will remain in the logical meter.  Only the soc, sop, capacity, etc will go to the battery pool.

ela-kotulska-frequenz replied:
> the battery power formula will remain in the logical meter. 

What if user would like to know active power for subset of batteries?

shsms replied:
Then they'll have to provide a custom formula.

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/logical_meter/_formula_generators.py:
```diff
@@ -91,3 +91,61 @@ async def grid_power_formula(
+        )
+
+        if not batteries:
+            raise RuntimeError("Unable to find any batteries in the component graph.")
```
I would add an exception like `ComponentNotFound` for these cases. I feel we are using too many generic exceptions so if at some point we need to start catching these exceptions and acting on them depending on the underlying issue, we'll have a **lot** of work converting all these `RuntimeError`s to something more meaningful all at once.

shsms replied:
also done.

shsms committed:
Remove unused `setup` method from formula engine tests

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Use custom exception types in the formula generator

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

ela-kotulska-frequenz commented on tests/microgrid/mock_api.py:
```diff
@@ -77,10 +78,22 @@ def __init__(
-        self, component_id: int, component_category: ComponentCategory.V
+        self,
+        component_id: int,
+        component_category: ComponentCategory.V,
+        inverter_type: InverterType.V = InverterType.TYPE_UNSPECIFIED,
```
`component_category` is expected as raw protobuf value.
`inverter_type` should be python Enum.
And the method creates struct of python Enums.

Maybe it would be more consistent if `inverter_type` would by expected as raw protobuf, too?

shsms replied:
it is the protobuf type already.  This is the mock server,  can't use our sdk types here.

ela-kotulska-frequenz commented on tests/microgrid/mock_api.py:
```diff
@@ -77,10 +78,22 @@ def __init__(
+        self,
+        component_id: int,
+        component_category: ComponentCategory.V,
+        inverter_type: InverterType.V = InverterType.TYPE_UNSPECIFIED,
     ) -> None:
         """Add a component to the mock service."""
-        self._components.append(Component(id=component_id, category=component_category))
+        if component_category == ComponentCategory.COMPONENT_CATEGORY_INVERTER:
+            self._components.append(
+                Component(
+                    id=component_id, category=component_category, inverter=inverter_type
+                )
+            )
+        else:
+            self._components.append(
+                Component(id=component_id, category=component_category)
+            )
```
If `inverter_type` would be renamed to `component_type` and it would be raw protobuf (as mentioned above), then you will be able to remove all `if` here and use  `_component_type_from_protobuf` (that you implemented in the first commit).

shsms replied:
No,  the `inverter` field in `Component` is a `oneof` defined [here](https://github.com/frequenz-floss/frequenz-api-microgrid/blob/v0.x.x/proto/frequenz/api/microgrid/microgrid.proto#L363-L369).  We can set it only if it is an inverter.  Else it will overwrite the value we set for a `battery` for example.  So it has to be done this way.

ela-kotulska-frequenz commented on src/frequenz/sdk/timeseries/logical_meter/_formula_generators.py:
```diff
@@ -0,0 +1,93 @@
+                "Unable to find a GRID component from the component graph."
+            )
+
+        grid_successors = component_graph.successors(grid_component.component_id)
```
But if umg_meter will be by load side, then grid_successor will have only one element: JUNCTION. And then formula below won't work.

shsms replied:
No,  we won't have JUNCTION components anymore,  components will just be directly connected.  This is be the component graph format for all locations that get added to the UI.  And v0.13.0 or higher of the sdk will not be deployed to locations that are not added to the UI.

ela-kotulska-frequenz commented on tests/timeseries/test_logical_meter.py:
```diff
@@ -130,22 +127,178 @@ def next_msg(value: float) -> ComponentData:
+        )
+        servicer.add_component(
+            7, microgrid_pb2.ComponentCategory.COMPONENT_CATEGORY_METER
+        )
```
Why there is no JUNCTION in this graph?

shsms replied:
Junction components are going away.

ela-kotulska-frequenz commented on src/frequenz/sdk/timeseries/logical_meter/_formula_generators/_battery_power_formula.py:
```diff
@@ -0,0 +1,69 @@
+
+        if len(batteries) != len(battery_inverters):
+            raise RuntimeError("Can't find inverters for all batteries.")
+
```
Why not to look for ComponentCategory.INVERTER?
And then iterate once over returned list to find inverters with ComponentType.BATTERY?

shsms replied:
They are both the same right?  Do you think it is better to start from the inverters for some reason?

ela-kotulska-frequenz replied:
I am telling that you don't need batteries, here. :) 
```
inverters = list(
            comp
            for comp in component_graph.components()
            if comp.category == ComponentCategory.INVERTER and comp.type == InverterType.BATTERY
        )
for idx, comp in enumerate(battery_inverters):
   ....
```

shsms replied:
oh that's true.  will update.

shsms replied:
done

shsms committed:
Add a formula generator for PV power

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms commented:
I will add some more tests as well, to this PR.

shsms committed:
Add flexible `MockMicrogrid` for testing logical meter formulas

... and update the logical meter tests to use a generated grid power
formula instead.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add tests for battery_power and pv_power formulas

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms commented:
This is ready for review again.  Sorry about the large number of changes.

ela-kotulska-frequenz commented on src/frequenz/sdk/timeseries/logical_meter/_formula_generators/_battery_power_formula.py:
```diff
@@ -38,22 +44,26 @@ async def generate(
 
         if len(batteries) != len(battery_inverters):
-            raise RuntimeError("Can't find inverters for all batteries.")
+            raise ComponentNotFound("Can't find inverters for all batteries.")
```
Don't you think it should be handled in ComponentGraph?

shsms replied:
this is now obsolete,  after https://github.com/frequenz-floss/frequenz-sdk-python/pull/103#discussion_r1050667778

ela-kotulska-frequenz commented on src/frequenz/sdk/timeseries/logical_meter/_formula_generators/_battery_power_formula.py:
```diff
@@ -0,0 +1,58 @@
+            if comp.category != ComponentCategory.INVERTER:
+                raise RuntimeError(
+                    f"Found a battery predecessor that's not an inverter: {comp}"
+                )
```
It should be removed, now :)

shsms replied:
oh wow 🤦🏽 done

leandro-lucarella-frequenz commented:
Ah, this looks very nice with the new exceptions and one class per formula :heart:
